validate_mission: add CNTO specifics to A3AA's validate mission

freghar opened the pull request:
This implements some of the checks described in https://github.com/CntoDev/cnto-additions/issues/28, see an example output (pieced together from failing checks), the first 4 of which come from A3AA directly (not implemented here):
```
Mission validation report
-------------------------
Overloaded units: PASS
Unique ACRE _ID_ radios: PASS
Playable unit ordering: PASS
Persistent callsigns: PASS
Game Type: FAIL

    Unexpected game type CTI.
    This may be okay for a custom mission, but none of the default modules
    use this type. If you're making a normal mission, check that a module
    composition was placed.

Respawn Type: FAIL

    Invalid respawn settings for Coop/PvP.
    Make sure the correct module composition was placed and respawn settings
    were not edited.

Respawn Markers Invisible: FAIL
 - marker:'respawn_west'

    Some respawn markers are visible.
    Double click them in the left panel (or on the map screen) and set their
    'Alpha' value all the way to 0%.

Respawn Markers Proximity: FAIL
 - marker:'respawn_west'

    One or more respawn markers is >100m away from nearest playable unit.
    This may be fine if you want players respawning in a location different
    from their starting one, but maybe you also accidentally moved playable
    units without moving their respawn marker.

Mission Info: FAIL

    Mission Title is empty.
    Mission Overview Picture is empty.
    Mission Overview Text is empty.
    Fill in at least the mission Title in Attributes -> General.

Changed CBA Settings: PASS

    ace_medical_statemachine_fatalinjuriesplayer = 0;
    ace_captives_allowhandcuffownside = false;
```

I'm not sure if the all-in-one-file approach is best - having checks in separate files would be cleaner, but it would also reduce copy-pasta-bility of some common code (you'd have to open / scour many files).

The "Mission Info" fails only if Title is empty, it will still report empty picture/text, but with a PASS as those are I think optional.

Comments?


PS: Note that this needs new A3AA (arma-additions) from Steam, currently (2021-04-22) not in the CNTO modset.

freghar committed:
validate_mission: add CNTO specifics to A3AA's validate mission

Signed-off-by: freghar <freghar@dummy.tld>

Seb105 commented on addons/validate_mission/fn_run.sqf:
```diff
@@ -0,0 +1,143 @@
+    };
+};
+
+private _check_respawn_marker_proximity = {
```
Maybe recommend/check for a "move on JIP" module to determine whether to pass/fail.

Respawn marker far away from players + move on JIP -> Probably intentional, pass
Respawn far away from players + no move on jip -> Maybe an accident, maybe MM isn't aware that JIPs don't spawn on respawn so fail + advise on a respawn marker move or to create a move on JIP module.

freghar replied:
Yeah, good point. I could also check the proximity of "Teleport on JIP" to the respawn marker, but maybe that's not worth it.

Seb105 reviewed (commented):
All features work for me, 1 comment about improving respawn checking.

freghar commented on addons/validate_mission/fn_run.sqf:
```diff
@@ -0,0 +1,143 @@
+        };
+        private _nearest = 100000;
+        {
+            _nearest = _nearest min (_x distance _marker_pos);
```
Note to self: `distance` between Position2D and Position3D assumes 0 in Z of the Position2D, so this check might fail on extreme slopes. Do `_x select [0,2]`.

freghar commented:
Rewrote the respawn marker proximity, added Teleport on JIP check, added respawn marker count check:

```
Respawn markers proximity: FAIL
 - marker:'respawn_west'

    One or more respawn markers is >100m away from nearest playable unit.
    Perhaps you accidentally moved playable units without moving their
    respawn marker?
    If the starting position is vastly different from the respawn location,
    consider using the 'Teleport on JIP' module.

// with Teleport on JIP placed:

Respawn Markers Proximity: FAIL
 - marker:'respawn_west'

    One or more respawn markers is >100m away from nearest playable unit.
    Perhaps you also accidentally moved playable units without moving their
    respawn marker?
    The 'Teleport on JIP' module is also >100m away from nearest respawn
    which is probably not intentional (JIPs joining on a place different
    from where players respawn). Move it closer to a respawn marker.

// count checker:

Respawn marker count: FAIL

    Found >1 (2) respawn markers.
    Using multiple respawn markers for Coop results in players respawning
    on randomly chosen respawn markers. This is a very rare use case.
    It's more likely that you deleted a player faction composition in 3D
    and re-placed it, leaving old map markers behind.
    Either delete those markers from the left panel (or map view) or simply
    delete the entire placed composition in map view + re-place it.
```


New commits:
```
add check for respawn marker count to validate_mission

redesign respawn marker proximity check for validate_mission

    The engine treats 'respawn_<side>' really only as priority hints,
    a non-west unit can easily respawn at 'respawn_west' if no other
    respawn marker is present, so we can't even catch typos.

    Add a 'Teleport on JIP' module check, make sure it's close enough
    to a respawn marker.
```

Seb105 reviewed (approved):
LGTM.

freghar committed:
redesign respawn marker proximity check for validate_mission

The engine treats 'respawn_<side>' really only as priority hints,
a non-west unit can easily respawn at 'respawn_west' if no other
respawn marker is present, so we can't even catch typos.

Add a 'Teleport on JIP' module check, make sure it's close enough
to a respawn marker.

Signed-off-by: freghar <freghar@dummy.tld>

freghar committed:
add check for respawn marker count to validate_mission

Signed-off-by: freghar <freghar@dummy.tld>
